### ISSUE TITLE
GraalVM on ARMv7 32 bits

### ISSUE THREAD

**Issue by aygalinc** [oracle/graal#1329]
Is it in the roadmap to enable Graal as JIT in ARMv7 32 bits Architecture or as native image generator ?

**Comment by bobmcwhirter** [comment]
I’m just a contributor but this is something I’d like to see and work towards. Dunno how ORCL feels or if it’s on any public or private roadmap though.

**Comment by vjovanov** [comment]
For native-image, we could use the LLVM backend to achieve this. 

This would not be as simple as telling LLVM to produce 32-bit code as there are assumptions in the code base related to the 64-bit word length. Nevertheless, using LLVM might reduce the amount of work significantly as those assumptions need to be dealt with anyhow.

**Comment by christianwimmer** [comment]
In theory, all of the Substrate VM code is written in a bit-width-agnostic way. For example, the `Word` type is automatically selected to be either a 32-bit or 64-bit primitive based on the architecture.
In practice, we have not had any 32-bit platform yet. So I would expect a few bugs to show up, but no major changes in the code based should be necessary.

As Vojin said, the LLVM backend is the easiest playground to experiment with that. We do not have it on any roadmap, but we will be happy to accept patches that make it work.

**Comment by carbotaniuman** [comment]
As someone who wants to contribute to this, what parts of the codebase needs to be extended to accomplish JIT on ARMv7?

**Comment by dmlloyd** [comment]
I believe the first step would likely be adding support to JVMCI.

**Comment by carbotaniuman** [comment]
I'll look into that. Do you know where that is? I've looked around Graal/Compiler for about 30 minutes now and haven't found the JVMCI portion.

**Comment by ezzarghili** [comment]
The JVMCI code base can be found here https://github.com/graalvm/graal-jvmci-8

**Comment by dougxc** [comment]
The above is the JDK 8 implementation of JVMCI which currently does not have an ARM support at all. It may be easier to start with https://github.com/openjdk/jdk which already has AArch64 support in JVMCI:

https://github.com/openjdk/jdk/tree/master/src/jdk.internal.vm.ci/share/classes/jdk.vm.ci.aarch64/src/jdk/vm/ci/aarch64
https://github.com/openjdk/jdk/tree/master/src/jdk.internal.vm.ci/share/classes/jdk.vm.ci.hotspot.aarch64/src/jdk/vm/ci/hotspot/aarch64
https://github.com/openjdk/jdk/blob/master/src/hotspot/cpu/aarch64/jvmciCodeInstaller_aarch64.cpp

**Comment by carbotaniuman** [comment]
Is there documentation for this? Such as whether the registers on Architecture should be 16 128-bit registers/32 64-bit registers or all the flags for the EnumSet?

**Comment by dmlloyd** [comment]
I haven't found any documentation on the JVMCI side, but I believe that 32-bit ARM (in normal operation) uses 16 32-bit registers, one of which (R15) is the PC and one which (R14) is usually used as the LR.  Also I think it's pretty normal to use R13 as the stack pointer.  There's more info at https://developer.arm.com/docs/100878/latest/registers-at-aarch32.

**Comment by carbotaniuman** [comment]
Yeah, I've read through that and I am reading through the ARMv7 Developer guide. I'm wondering if the NEON registers should be the 128-bit or 64 bit views and what flags exist in VMVersion (or do I make those myself?).

**Comment by adinn** [comment]
Hi carbotaniuman.

Are you aware that armv7 is essentially a completely different ISA to armv8. In consequence, your implementation will be unable to achieve significant reuse of the Graal aarch64 components in the compiler and substratevm trees. In other words, if you want to run graal on armv7 you not only need an JVMCI-enabled OpenJDK runtime for armv7 to bootstrap it with (e.g. the jdk dev tree that Doug pointed you at) but you also need to implement *from scratch* armv7 equivalents of all the x86- or aarch64-specific code in the compiler and substratevm directories. That's quite a tall order!

**Comment by bobmcwhirter** [comment]
@carbotaniuman But also, realize that @adinn has coworkers who might also pitch in, such as myself or @dmlloyd or anyone else interested in PiZeros or other hardware. :)

That makes everyone's portion 1/3 of a tall order, which starts to become only a slight more than average order.

**Comment by adinn** [comment]
Hi Bob,

I'm not wanting to stop anyone having fun here -- just pointing out exactly how much fun you face :-)

**Comment by kg6zvp** [comment]
Really would like to be able to compile AOT for ARMv7 32-bit systems like Raspberry Pi Zero W!

I posted a bounty for this on BountySource. It's not much, but if everyone pitches in a few dollars I'm sure someone will take the time to get it done: https://www.bountysource.com/issues/74587879-graalvm-on-armv7-32-bits

**Comment by The-Alchemist** [comment]
> Really would like to be able to compile AOT for ARMv7 32-bit systems like Raspberry Pi Zero W!
> 
> I posted a bounty for this on BountySource. It's not much, but if everyone pitches in a few dollars I'm sure someone will take the time to get it done: https://www.bountysource.com/issues/74587879-graalvm-on-armv7-32-bits

Pitched in a few more bucks!

**Comment by kg6zvp** [comment]
Added another $100 to the bounty. We need AOT for Raspberry Pi Zero Java projects! Anyone even thinking about picking this up?

**Comment by jabrena** [comment]
It could be fine to compile for ARM32 and run in this kind of devices:
https://github.com/numworks/epsilon

**Comment by jwogit** [comment]
GraalVM for ARMv6 + v7 would be great. It would be a precious contribution to sustainability. There are thousands of flawless PI's 1st and 2nd generation out there and you would give them another chance by this.

**Comment by koutheir** [comment]
I think the Raspberry Pi 1 runs an ARMv6K, which is close enough to ARMv7-A. Adapting an ARMv7-A port to ARMv6K should be fairly simple.

**Comment by jason-lindholm** [comment]
+1

**Comment by ericCHensssssssss** [comment]
+1

**Comment by etgundam53** [comment]
+1

**Comment by ThatMG393** [comment]
+1

**Comment by ericCHensssssssss** [comment]
+1 +1 +1

**Comment by ancienter** [comment]
+1 +1 +1 +1

**Comment by brunocaballero** [comment]
MicroDoc offers a port of Native Image for 32-bit platforms, including ARMv7-A and x86. With this solution, you can generate images on a Linux x86_64 system that are compatible with 32-bit platforms. For more details, visit: [MicroDoc GraalVM](https://www.microdoc.com/graal-vm/)

**Comment by ThatMG393** [comment]
+1

**Comment by jabrena** [comment]
One interesting use case:

- Calculators: https://www.numworks.com/engineering/hardware/electrical/parts/

```
MCU
STM32F730V8T6
This is the brain and heart of our calculator. This chip combines an ARMv7-M Cortex-M7 core clocked at 216 MHz and 256K of Static RAM.
```

**Comment by dwalend** [comment]
Also, Lego Ev3. https://www.ev3dev.org/docs/getting-started/

**Comment by tkeksa** [comment]
And many NXP iMX7 based boards.

**Comment by Ramblurr** [comment]
Another use case: Nearly all commercially available e-ink e-reader devices run on armv7l.
